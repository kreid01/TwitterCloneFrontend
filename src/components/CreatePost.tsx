import React from 'react' 
import { FaImage, FaCalendar, FaSmile} from 'react-icons/fa'
import {IconBaseProps} from 'react-icons'

interface newPost {
    postTextBody: string
    postMedia: string
}

interface Props {
    newPost: newPost
    handleChange: (event: React.ChangeEvent<HTMLTextAreaElement>) => void
}

export const CreatePost: React.FC<Props> = ({ handleChange }, props: Props) => {
    

    
    return (
        <div className='bottom-3 px-3 mb-36 flex'>
            <img src='https://cdn-icons-png.flaticon.com/512/149/149071.png' className='w-12 h-12'alt=''/>
            <div>
                <textarea 
                onChange={(event) => handleChange(event)}
                name='postTextBody'
                className="pl-3 min-w-28 border-gray-300 resize-none text-gray-900 text-sm rounded-lg outline-0 block p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white" placeholder="Whats Happening?" required/>
                <div className='flex'>
                    <NewPostButtons icon={<FaImage/>}/>
                    <NewPostButtons icon={<FaCalendar/>}/>
                    <NewPostButtons icon={<FaSmile/>}/>
                    <button
                     className='font-bold text-white ml-auto bg-blue-500 hover:bg-blue-400 p-2 pl-4 pr-4 rounded-2xl'>Tweet</button>
                </div>
            </div>   
        </div>
    )

}

const NewPostButtons = ({ icon }: {icon: IconBaseProps})  => (
    <div className='post-icon'>
        <>{icon}</>
    </div>
)