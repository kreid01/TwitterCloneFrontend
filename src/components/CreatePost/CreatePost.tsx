import React, { useState, useEffect } from 'react' 
import {FaCalendar, FaSmile, FaImage} from 'react-icons/fa'
import {IconBaseProps} from 'react-icons'
import { TweetButton } from '../TweetButton/TweetButton'
import { FileUpload } from './FileUpload'

interface newPost {
    postTextBody: string
    postMedia: string
}

interface Props {
    newPost: newPost
    setNewPost: React.Dispatch<React.SetStateAction<{
        postTextBody: string;
        postMedia: string;
        userAt: string;
        userName: string;
        userImg: string;
    }>>
    handleChange: (event: React.ChangeEvent<HTMLTextAreaElement>) => void
    handleTweet: () => void
}

export const CreatePost: React.FC<Props> = ({ handleChange, handleTweet, setNewPost }, props: Props) => {
 
    const [image, setImage] = useState<File | null>()
    const [preview, setPreview] = useState<string | null>()

    const handleFileUpload = (event: React.ChangeEvent<HTMLInputElement>) => {
        const file = event.target.files
        if(file) {
            if(file) {
            setImage(file[0])
            } else {
                setImage(null)
            }     
        }
    } 

    useEffect(() => {
        if(image) {
            const reader = new FileReader()
            let url = ''
            reader.onloadend = () => {
                setPreview(reader.result as string)
                setNewPost(prevState => ({
                    ...prevState, postMedia: url
                }))
            }
            reader.readAsDataURL(image)
        } else {
            setPreview(null)
        } 
    },[image])
 
   


    return (
        
        <div className='bottom-3 px-3 mb-18 flex'>
            <img src='https://cdn-icons-png.flaticon.com/512/149/149071.png' className='w-12 h-12'alt=''/>
            <div>
            <p>{preview}</p>
                <form>
                    <textarea 
                    onChange={(event) => handleChange(event)}
                    name='postTextBody'
                    className="text-lg pl-3 w-72 border-gray-300 h-36 resize-none
                    text-gray-900 rounded-lg outline-0 block p-2.5 dark:bg-gray-700
                    dark:border-gray-600 dark:placeholder-gray-400 dark:text-white"
                    placeholder="Whats Happening?" required/>
                    <img alt='' className='w-80 pt-2 pb-6 max-h-72 rounded-xl' src={preview as string}/>
                    <div className='flex'>
                        <FileUpload 
                        label={<NewPostButtons icon={<FaImage />}/>}
                        handleFileUpload={handleFileUpload}/>
                        <NewPostButtons icon={<FaCalendar/>}/>
                        <NewPostButtons icon={<FaSmile/>}/>
                        <TweetButton
                        label="Tweet"
                         handleTweet={handleTweet}></TweetButton>
                    </div>
                </form>
            </div>   
        </div>
    )

}

export const NewPostButtons = ({ icon }: {icon: IconBaseProps})  => (
    <div className='post-icon'>
        <>{icon}</>
    </div>
)