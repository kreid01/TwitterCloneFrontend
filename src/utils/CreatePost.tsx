import React, { useState, useEffect } from 'react' 
import { FaImage, FaCalendar, FaSmile} from 'react-icons/fa'
import {IconBaseProps} from 'react-icons'

interface newPost {
    postTextBody: string
    postMedia: string
}

interface Props {
    newPost: newPost
    handleChange: (event: React.ChangeEvent<HTMLTextAreaElement>) => void
    handleTweet: () => void
}

export const CreatePost: React.FC<Props> = ({ handleChange, handleTweet}, props: Props) => {
 
    const [image, setImage] = useState<File | null>()
    const [preview, setPreview] = useState<string | null>()

    const handleFileUpload = (event: React.ChangeEvent<HTMLInputElement>) => {
        const file = event.target.files
        if(file) {
            if(file) {
            setImage(file[0])
            } else {
                setImage(null)
            }     
        }
    } 

    useEffect(() => {
        if(image) {
            const reader = new FileReader()
            reader.onloadend = () => {
                setPreview(reader.result as string)
            }
            reader.readAsDataURL(image)
            console.log(preview)
        } else {
            setPreview(null)
        } 
    },[image])

    
    return (
        
        <div className='bottom-3 px-3 mb-18 flex'>
            <img src='https://cdn-icons-png.flaticon.com/512/149/149071.png' className='w-12 h-12'alt=''/>
            <div>
            <p>{preview}</p>
                <textarea 
                onChange={(event) => handleChange(event)}
                name='postTextBody'
                className="text-lg pl-3 w-72 border-gray-300 h-36 resize-none
                 text-gray-900 rounded-lg outline-0 block p-2.5 dark:bg-gray-700
                  dark:border-gray-600 dark:placeholder-gray-400 dark:text-white"
                   placeholder="Whats Happening?" required/>
                <img alt='' className='w-80 pt-2 pb-6 max-h-72 rounded-xl' src={preview as string}/>
                <div className='flex'>
                    <input 
                    accept="image/*"
                    className='hidden'
                    name='postMedia'
                    id='postMedia' onChange={(event) => handleFileUpload(event)} type='file'/>
                    <label htmlFor='postMedia'><NewPostButtons icon={<FaImage />}/></label>
                    <NewPostButtons icon={<FaCalendar/>}/>
                    <NewPostButtons icon={<FaSmile/>}/>
                    <button
                    type='submit'
                    onClick={handleTweet}
                     className='font-bold text-white ml-auto bg-blue-500 hover:bg-blue-400 p-2 pl-4 pr-4 rounded-2xl'>Tweet</button>
                </div>
            </div>   
        </div>
    )

}

const NewPostButtons = ({ icon }: {icon: IconBaseProps})  => (
    <div className='post-icon'>
        <>{icon}</>
    </div>
)