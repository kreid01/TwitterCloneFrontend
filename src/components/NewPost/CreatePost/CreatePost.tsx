import React, { useState, useEffect } from "react";
import { FaCalendar, FaSmile } from "react-icons/fa";
import { IconBaseProps } from "react-icons";
import { TweetButton } from "../TweetButton/TweetButton";
import { FileUpload } from "../FileUpload/FileUpload";
import { PreviewImg } from "../PreviewImg/PreviewImg";
import { UserTextInput } from "../UserTextInput/UserTextInput";
import { useFileReader } from "../../../hooks/utils/useFileReader";
import { IPost } from "../../../consts/Interface";
import { postPost } from "../../../services/posts/postPost";
import { useGetUser } from "context/UserContext";

interface Props {
  setPosts: React.Dispatch<React.SetStateAction<IPost[] | undefined>>;
}

export const CreatePost: React.FC<Props> = ({ setPosts }) => {
  const user = useGetUser();
  const newPostInitialState = {
    postTextBody: "",
    postMedia: "",
    posterId: user?.userId,
    userAt: user?.userAt,
    userName: user?.userName,
    userImg: user?.userImg,
    commentCount: 0,
    retweetCount: 0,
    likeCount: 0,
    postDate: "",
    id: 0,
  };
  const [newPost, setNewPost] = useState(newPostInitialState);
  const [file, setFile] = useState<FileList | null>();
  const handleFileUpload = (event: React.ChangeEvent<HTMLInputElement>) => {
    setFile(event.target.files as FileList);
  };
  const { preview, fileFromReader, setPreview } = useFileReader(
    file as FileList
  );
  const handleChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
    setNewPost((prevState) => ({
      ...prevState,
      [event.target.name]: event.target.value,
    }));
  };

  const resetNewPost = () => {
    setNewPost(newPostInitialState);
  };

  const handleTweet = () => {
    postPost(newPost);
    setPosts((prevArr) => [...(prevArr as Array<IPost>), newPost]);
    resetNewPost();
    setPreview(null);
  };

  useEffect(() => {
    setNewPost((prevState) => ({
      ...prevState,
      postMedia: fileFromReader as string,
    }));
  }, [fileFromReader]);

  return (
    <div className="bottom-3 px-3 mb-6 flex">
      <img
        src="https://cdn-icons-png.flaticon.com/512/149/149071.png"
        className="w-12 h-12"
        alt=""
      />
      <div>
        <UserTextInput
          value={newPost.postTextBody}
          name="postTextBody"
          handleChange={handleChange}
          placeholder="What's Happening?"
        />
        <PreviewImg imgSrc={preview as string} />
        <div className="flex">
          <FileUpload name="postMedia" handleFileUpload={handleFileUpload} />
          <NewPostButtons icon={<FaCalendar />} />
          <NewPostButtons icon={<FaSmile />} />
          <TweetButton
            label="Tweet"
            handleTweet={handleTweet}
            isDisabled={newPost.postTextBody.length === 0 ? true : false}
          ></TweetButton>
        </div>
      </div>
    </div>
  );
};

export const NewPostButtons = ({ icon }: { icon: IconBaseProps }) => (
  <div className="post-icon">
    <>{icon}</>
  </div>
);
