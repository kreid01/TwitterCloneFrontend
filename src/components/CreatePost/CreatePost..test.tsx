import React from 'react'
import { CreatePost } from './CreatePost'

import { cleanup, render } from '@testing-library/react';
import "@testing-library/jest-dom";

import renderer from 'react-test-renderer'

const handleTweet = () => {}
const setNewPost = () => {}
const handleFileUploadMockFn = jest.fn()

afterEach(cleanup)
it("matched snapshot", () => {
    const tree = renderer.create(
        <CreatePost
        newPost={{
            postTextBody:'',
            postMedia: '' }}
        handleTweet={handleTweet}
        handleChange={handleFileUploadMockFn}
        setNewPost={setNewPost}/>
      ).toJSON()
      expect(tree).toMatchSnapshot()     
})

it("renders without crashing", () => {
    render(
        <CreatePost
        newPost={{
            postTextBody:'',
            postMedia: '' }}
        handleTweet={handleTweet}
        handleChange={handleFileUploadMockFn}
        setNewPost={setNewPost}/>)   
})
