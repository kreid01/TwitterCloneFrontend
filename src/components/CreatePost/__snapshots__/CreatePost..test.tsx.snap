// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matched snapshot 1`] = `
<div
  className="bottom-3 px-3 mb-18 flex"
>
  <img
    alt=""
    className="w-12 h-12"
    src="https://cdn-icons-png.flaticon.com/512/149/149071.png"
  />
  <div>
    <p />
    <form>
      <textarea
        className="text-lg pl-3 w-72 border-gray-300 h-36 resize-none
 text-gray-900 rounded-lg outline-0 block p-2.5 dark:bg-gray-700
 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white"
        name="postTextBody"
        onChange={[Function]}
        placeholder="Whats Happening?"
        required={true}
      />
      <img
        alt=""
        className="w-80 pt-2 pb-6 max-h-72 rounded-xl"
      />
      <div
        className="flex"
      >
        <input
          accept="image/*"
          className="hidden"
          id="postMedia"
          name="postMedia"
          onChange={[Function]}
          type="file"
        />
        <label
          htmlFor="postMedia"
        >
          <div
            className="post-icon"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M464 448H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h416c26.51 0 48 21.49 48 48v288c0 26.51-21.49 48-48 48zM112 120c-30.928 0-56 25.072-56 56s25.072 56 56 56 56-25.072 56-56-25.072-56-56-56zM64 384h384V272l-87.515-87.515c-4.686-4.686-12.284-4.686-16.971 0L208 320l-55.515-55.515c-4.686-4.686-12.284-4.686-16.971 0L64 336v48z"
              />
            </svg>
          </div>
        </label>
        <div
          className="post-icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 192h424c6.6 0 12 5.4 12 12v260c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V204c0-6.6 5.4-12 12-12zm436-44v-36c0-26.5-21.5-48-48-48h-48V12c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v52H160V12c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v52H48C21.5 64 0 85.5 0 112v36c0 6.6 5.4 12 12 12h424c6.6 0 12-5.4 12-12z"
            />
          </svg>
        </div>
        <div
          className="post-icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 496 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm80 168c17.7 0 32 14.3 32 32s-14.3 32-32 32-32-14.3-32-32 14.3-32 32-32zm-160 0c17.7 0 32 14.3 32 32s-14.3 32-32 32-32-14.3-32-32 14.3-32 32-32zm194.8 170.2C334.3 380.4 292.5 400 248 400s-86.3-19.6-114.8-53.8c-13.6-16.3 11-36.7 24.6-20.5 22.4 26.9 55.2 42.2 90.2 42.2s67.8-15.4 90.2-42.2c13.4-16.2 38.1 4.2 24.6 20.5z"
            />
          </svg>
        </div>
        <button
          className="font-bold text-white ml-auto 
 bg-blue-500 hover:bg-blue-400 p-2 pl-4
 pr-4 rounded-2xl"
          data-testid="tweetButton"
          onClick={[Function]}
        >
          Tweet
        </button>
      </div>
    </form>
  </div>
</div>
`;
